components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaClient.kt

```diff
@@ -0,0 +1,423 @@
+                coroutineContext,
+            )
+        }
+    }
+
+    private fun tryPersistState() {
+        @Suppress("TooGenericExceptionCaught")
+        try {
+            persistCallback.persist(inner.toJson())
+        } catch (e: Exception) {
+            Log.e(LOG_TAG, "Error saving the FirefoxAccount state.")
+        }
+    }
+
+    /**
+     * Queue an FxaAction for processing and immediately return
+     *
+     * Use the FxaEventHandler passed to the constructor to respond to the results of queued
+     * actions.
+     *
+     * Why do this?
+     *   - Actions are processed serially.  For example, there's no chance of CompleteOAuthFlow and
+     *     Disconnect being executed at the same time from different threads.
+     *   - Application events are also sent serially. If one action causes an [AUTH_CHECK_STARTED]
+     *     state transition change, then the next causes [FxaAuthState.Disconnected], the
+     *     [FxaEventHandler.onStateChange] callback for the second change won't be called until
+     *     after the callback for the first change returns. This allows applications to ensure that
+     *     their UI reflects the current state.
+     *   - Actions are retried in the face of network errors and checkAuthorizationStatus is called on
+     *     authorization errors.  This allows the Rust client to recover when possible, for example
+     *     from expired access tokens when it holds a valid refresh token.
+     */
+    fun queueAction(action: FxaAction) {
+        actionProcessor.queue(action)
+    }
+
+    // This handles queueAction for us
+    private val actionProcessor = FxaActionProcessor(inner, eventHandler, { tryPersistState() })
+    init {
+        runActionProcessorManager(actionProcessor, coroutineContext)
+    }
+
+    /**
+     * Get the current authentication state
+     *
+     * FIXME: https://github.com/mozilla/application-services/issues/5819
+     */
+    fun getAuthState(): FxaAuthState = FxaAuthState.fromRust(inner.getAuthState())
+
+    // Wraps method calls that don't change the state, like [getAccessToken]
+    private suspend fun<T> wrapMethodCall(name: String, methodCall: () -> T): T {
+        Log.d(LOG_TAG, "Running: $name")
+        return withContext(coroutineContext) {
+            try {
+                methodCall()
+            } catch (e: FxaException.Authentication) {
+                queueAction(FxaAction.CheckAuthorization)
+                throw e
+            }
+        }
+    }
+
+    // Does what wrapMethodCall does and also ensures that the state is perisisted at the end of the
+    // operation.
+    private suspend fun<T> wrapMethodCallAndPersist(name: String, methodCall: () -> T): T {
+        return try {
+            wrapMethodCall(name, methodCall)
+        } finally {
+            // Use a finally block since we want to perisist the state regardless of if the method
+            // succeeded or not.
+            tryPersistState()
```

Comment:
I wrote it up this way because that's how the existing code worked ([for example, getProfile()](https://github.com/mozilla/application-services/blob/main/components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/PersistedFirefoxAccount.kt#L152-L158)).  I think that's intended to address the case where you the client tries to fetch the profile, realizes its auth token is bad so it deletes it, and returns an auth error.  In that case you still want to persist the state even though the action failed.

So I think this should be okay, since that's how the current API works and I don't think we've ever wanted to break it up into smaller chunks.

However, your idea of a transactional-style API is really interesting.  In addition to allowing us to expose smaller operations, it could also handle the serialization issue.  The whole idea of the action / event system is to ensure that we process in the correct order and also that the consumer responds to the results in the correct order.   Another approach would be to have the consumer start a transaction, perform some operations, handle the results, then drop the transaction.  If we enforced only 1 transaction running at a time, then I think this achieves the same result and maybe would be a nicer API to work with.